src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -81,37 +99,114 @@ private String decodeAlarm(String event) {
                 return Position.ALARM_GEOFENCE_EXIT;
             case "6":
                 return Position.ALARM_OVERSPEED;
+            case "5":
+                return Position.ALARM_POWER_CUT;
             case "3":
                 return Position.ALARM_SOS;
             default:
                 return null;
         }
     }
 
-    @Override
-    protected Object decode(
-            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+    private String decodeEvent(String event, Position position) {
 
-        String sentence = (String) msg;
+        if (event.length() == 1) {
+            char inputStatus = event.charAt(0);
+            if (inputStatus >= 'A' && inputStatus <= 'D') {
+                int inputStatusInt = inputStatus - 'A';
+                position.set(Position.PREFIX_IN + 1, inputStatusInt & 1);
+                position.set(Position.PREFIX_IN + 2, inputStatusInt & 2);
+                position.set(Position.KEY_IGNITION, ((inputStatusInt & 1) != 0) ? true : false);
+                return null;
+            }
+        }
 
-        if (sentence.startsWith("$ECHK") && channel != null) {
-            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress)); // heartbeat
-            return null;
+        return event;
+    }
+
+    private String getDevicePassword(DeviceSession deviceSession) {
```

Comment:
This like code duplication with what we have in the encoder.

Reply:
Good. I'll fix it. Thanks.

Reply:
Sorry, what is the fix here?

Reply:
Sorry, there is something I'm missing. We don't have code like this anymore in the encoder. Thank you.

Reply:
Isn't in the base encoder?

Reply:
Are you talking about the `initDevicePassword` method on `BaseProtocolEncoder` class? The one that is protected and receives a `Command` object that I don't need here?. It can't be accessed directly from the decoder. Maybe we can put both methods on the base encoder, unify shared code and provide a public method to get the password directly from a `Device` object or a device ID string. What do you propose? Thanks.

Reply:
Yes, we need to unify the code. One option is to define it in the `IdentityManager`.

Reply:
Great idea. I'll work on it. Thanks...